Set `transition-behavior` in a constructed stylesheet

DmitrySharabin opened the pull request:
- Get rid of all useless (invalid) properties in the browser dev tools. Appeared after https://github.com/LeaVerou/style-observer/commit/e5de44d78269ded5bdd34efdde0d7a3b220aa3fd
- Allow adopting CSS in shadow roots.

LeaVerou commented on src/util/adopt-css.js:
```diff
@@ -4,7 +4,7 @@ let style;
  */
 export default function adoptCSS (css, document = globalThis.document) {
-	let window = document.defaultView;
+	let window = document.defaultView ?? document.ownerDocument?.defaultView;
```
I don't think `ownerDocument` on `document` makes a lot of sense, wouldn't it always point to itself? Oh, I see, it's a local variable. We should rename it to something else, like `root`.

DmitrySharabin replied:
Done

LeaVerou commented on src/element-style-observer.js:
```diff
@@ -96,17 +97,40 @@ export default class ElementStyleObserver {
 		this.updateTransition({ firstTime });
 
+		// Initialize transition behavior for the main document
+		this.#setTransitionBehavior(this.target.ownerDocument);
```
I don't think this is necessary: It will be added once we observe a node in the light DOM, and if we never do, then we don't need it. Additionally, it's not desirable when Style-Observer is used within components, as they don't typically want to leak stuff to the light DOM unless absolutely necessary.

DmitrySharabin replied:
Got it. Thanks!

LeaVerou commented on src/element-style-observer.js:
```diff
@@ -96,17 +97,40 @@ export default class ElementStyleObserver {
+
+		// If the target is in a shadow root, initialize that too
+		let root = this.target.getRootNode();
+		if (root instanceof ShadowRoot) {
```
This check is not needed, see above.

LeaVerou commented on src/element-style-observer.js:
```diff
@@ -96,17 +97,40 @@ export default class ElementStyleObserver {
+		// Initialize transition behavior for the main document
+		this.#setTransitionBehavior(this.target.ownerDocument);
+
+		// If the target is in a shadow root, initialize that too
```
Is init the best place for this? What if the element is later moved to another shadow root or document? 

One way would be to cache the element's root node in a [`WeakRef`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef) and then do this both on init as well as on reparenting (which we know happened if  IntersectionObserver fires + the root node has changed)

LeaVerou reviewed (changes requested):
Great start, see comments!

DmitrySharabin commented on src/element-style-observer.js:
```diff
@@ -96,17 +97,40 @@ export default class ElementStyleObserver {
+	 * Set (ones per root) transition behavior on an element's root node (document or shadow root).
+	 * @param {Document|ShadowRoot} root - The root node to set transition behavior on.
+	 */
+	#setTransitionBehavior (root) {
```
Done

DmitrySharabin commented:
Thank you for your feedback and suggestions. I think I addressed all of them and am ready for another round.

LeaVerou commented on src/element-style-observer.js:
```diff
@@ -78,11 +85,18 @@ export default class ElementStyleObserver {
+			if (this.target.getRootNode() !== this.rootNode.deref()) {
+				// The element was moved to another document or shadow root
+				this.rootNode = new WeakRef(this.target.getRootNode());
+				this.#setTreeCSS(this.rootNode.deref());
```
We shouldn't need to dereference this again, setting the `rootNode` should suffice.

LeaVerou replied:
In fact, this whole check should probably be in `#setTreeCSS()`.

LeaVerou reviewed (changes requested):
Almost there! See comment.

DmitrySharabin commented:
As you suggested, I moved the logic to `#setTreeCSS()` and fixed a couple of bugs along the way: `node.getRootNode()` can return an element ([the root of the DOM tree the node belongs to](https://developer.mozilla.org/en-US/docs/Web/API/Node/getRootNode#return_value)), so we need to handle it properly.

I also made `nodeRoot` a private property since we only use it internally to store a weak ref to the target's root node. Authors can easily get the root node of an observed element with `element.getRootNode()` themselves if needed.

Could you please take another look?

DmitrySharabin commented on src/element-style-observer.js:
```diff
@@ -96,17 +106,41 @@ export default class ElementStyleObserver {
 		let firstTime = this.constructor.all.get(this.target).size === 1;
 		this.updateTransition({ firstTime });
 
+		this.#setTreeCSS(this.target.getRootNode());
```
> What I was saying is, you never call this function with any other argument. So just don't have any other arguments and do the `this.target.getRootNode()` _in_ the function. Then anything that may have changed the root node can just call it.

Oh, right. That makes more sense. I didn't even thing about it. Thanks!

DmitrySharabin replied:
> In fact, maybe make it a function in the public API and call it `targetMoved()`? (no strong opinion)

Let's do it. It might be useful.

DmitrySharabin commented on src/element-style-observer.js:
```diff
@@ -270,14 +282,33 @@ export default class ElementStyleObserver {
+	/**
+	 * Set (ones per root) CSS on an element's root node (document or shadow root).
+	 */
+	targetMoved () {
```
Will do
